File: clients/python/src/agenticdriver/_transport.py
"""Shared connection policy; execution inactivity is measured on the host."""

import math
from urllib.parse import urlsplit, urlunsplit

from ._errors import DriverError
from ._protocol import PROTOCOL_VERSION

TERMINAL_EVENTS = {"run.completed", "run.failed", "run.cancelled"}


def connection_url(url: str, token: str) -> str:
    try:
        if any(ord(c) <= 32 or ord(c) == 127 for c in url):
            raise ValueError()
        parsed = urlsplit(url)
        if (
            parsed.username
            or parsed.password
            or parsed.query
            or parsed.fragment
            or not parsed.hostname
            or parsed.port == 0
            or not (
                parsed.scheme == "https"
                or (
                    parsed.scheme == "http"
                    and parsed.hostname in {"localhost", "127.0.0.1", "::1"}
                )
            )
        ):
            raise ValueError()
    except (ValueError, TypeError):
        raise DriverError(
            "INSECURE_TRANSPORT",
            "Use HTTPS, or HTTP on loopback, without URL credentials, query, or fragment.",
        ) from None
    if not token or any(ord(c) < 33 or ord(c) > 126 for c in token):
        raise DriverError(
            "AUTH_REQUIRED",
            "A driver bearer token is required, without whitespace or control characters.",
        )
    return urlunsplit(parsed).rstrip("/") + "/"


def validate_timeout(value: float | None) -> float | None:
    if value is not None and (
        isinstance(value, bool) or not math.isfinite(value) or value <= 0
    ):
        raise ValueError(
            "An I/O timeout must be a finite positive number of seconds, or None."
        )
    return value


def headers(token: str, stream: bool, body: bool) -> dict[str, str]:
    result = {
        "Authorization": f"Bearer {token}",
        "Accept": "text/event-stream" if stream else "application/json",
        "AgenticDriver-Version": PROTOCOL_VERSION,
        "AgenticDriver-Accept-Optional-Events": "true",
    }
    if body:
        result["Content-Type"] = "application/json"
    return result


def check_version(versions: list[str] | None) -> None:
    if versions is not None and versions != [PROTOCOL_VERSION]:
        raise DriverError(
            "UNSUPPORTED_PROTOCOL_VERSION",
            "The host selected an unsupported wire protocol version.",
        )

File: clients/rust/src/transport.rs
use crate::{protocol_error, Error, Result, PROTOCOL_VERSION};

pub(crate) fn endpoint(url: &str, token: &str) -> Result<reqwest::Url> {
    let mut base = reqwest::Url::parse(url).map_err(|_| Error::Protocol("Invalid driver URL."))?;
    let local = matches!(
        base.host_str(),
        Some("localhost" | "127.0.0.1" | "[::1]" | "::1")
    );
    if !base.username().is_empty()
        || base.password().is_some()
        || base.query().is_some()
        || base.fragment().is_some()
        || base.host_str().is_none()
        || !(base.scheme() == "https" || (base.scheme() == "http" && local))
    {
        return Err(Error::Protocol(
            "Use HTTPS, or HTTP on loopback, without URL credentials, query, or fragment.",
        ));
    }
    if token.is_empty()
        || reqwest::header::HeaderValue::from_str(&format!("Bearer {token}")).is_err()
    {
        return Err(Error::Protocol("A valid driver bearer token is required."));
    }
    if !base.path().ends_with('/') {
        base.set_path(&format!("{}/", base.path()));
    }
    Ok(base)
}

pub(crate) fn check_version(headers: &reqwest::header::HeaderMap) -> Result<()> {
    if let Some(version) = headers.get("AgenticDriver-Version") {
        if headers.get_all("AgenticDriver-Version").iter().count() != 1
            || version.to_str().ok() != Some(PROTOCOL_VERSION)
        {
            return Err(protocol_error(
                "UNSUPPORTED_PROTOCOL_VERSION",
                "The host selected an unsupported wire protocol version.",
            ));
        }
    }
    Ok(())
}

pub(crate) fn check_sse(headers: &reqwest::header::HeaderMap) -> Result<()> {
    if headers
        .get("content-type")
        .and_then(|v| v.to_str().ok())
        .and_then(|v| v.split(';').next())
        .is_some_and(|v| v.trim().eq_ignore_ascii_case("text/event-stream"))
    {
        Ok(())
    } else {
        Err(protocol_error(
            "INVALID_RESPONSE",
            "Expected an SSE response.",
        ))
    }
}
